astropy/io/fits/column.py

```diff
@@ -1528,7 +1528,25 @@ def _init_from_array(self, array):
         for idx in range(len(array.dtype)):
             cname = array.dtype.names[idx]
             ftype = array.dtype.fields[cname][0]
-            format = self._col_format_cls.from_recformat(ftype)
+            len_array = len(array)
+            if ftype.kind in {"O", "V"} and ftype.hasobject:
+                if array.dtype[idx].subdtype is not None:
+                    len_array = array.dtype[idx].subdtype[1][0]
+                    dtypes = {
+                        np.array(array[cname][idx][i]).dtype for i in range(len_array)
+                    }
+                else:
+                    dtypes = {np.array(array[cname][i]).dtype for i in range(len_array)}
+                if (len(dtypes) > 1) or (np.dtype("O") in dtypes):
+                    raise TypeError(
+                        "Column contains not supported object types or "
+                        "different types are present in the same column."
```

Comment:
Maybe something like this to provide more information?
```suggestion
                        f"Column {cname} contains unsupported object types or "
                        f"mixed types: {dtypes}"
```